src/crimson/os/seastore/seastore_types.h

```diff
@@ -411,38 +412,123 @@ class segment_map_t {
  * Fresh extents during a transaction are refered to by
  * record_relative paddrs.
  */
+constexpr uint16_t DEV_ADDR_LEN_BITS = 64 - DEVICE_ID_LEN_BITS;
+enum class addr_types_t : uint8_t {
+  SEGMENT = 0,
+  RANDOM_BLOCK = 1
+};
+using common_addr_t = uint64_t;
+
 struct paddr_t {
-  segment_id_t segment = NULL_SEG_ID;
-  segment_off_t offset = NULL_SEG_OFF;
+  common_addr_t dev_addr = 0;
 
   paddr_t() = default;
-  paddr_t(device_id_t id, device_segment_id_t sgt, segment_off_t offset)
-    : segment(segment_id_t(id, sgt)), offset(offset) {}
+  constexpr paddr_t(common_addr_t addr) 
+    : dev_addr(addr) {}
+  paddr_t(device_id_t id, segment_id_t seg, segment_off_t offset);
 
-  constexpr paddr_t(segment_id_t segment, segment_off_t offset)
-    : segment(segment), offset(offset) {}
+  // use 1bit in device_id_t for address type
+  void set_device_id(device_id_t id, addr_types_t type = addr_types_t::SEGMENT) {
+    dev_addr |= ((common_addr_t)(id & 0x8) << DEV_ADDR_LEN_BITS);
+    dev_addr &= (common_addr_t)type << (std::numeric_limits<common_addr_t>::digits - 1);
+  }
+  device_id_t get_device_id() const {
+    return static_cast<device_id_t>((dev_addr >> DEV_ADDR_LEN_BITS) & 0xF);
+  }
+  addr_types_t get_addr_type() const {
+    return (addr_types_t)((dev_addr >> (std::numeric_limits<common_addr_t>::digits - 1)) & 1);
+  }
+
+  bool is_relative() const;
+  bool is_record_relative() const;
+  bool is_null() const;
+  bool is_block_relative() const;
+  bool is_zero() const;
+  bool is_real() const;
+
+  paddr_t add_offset(int32_t o) const;
+  paddr_t add_relative(paddr_t o) const;
+  paddr_t add_block_relative(paddr_t o) const;
+  paddr_t add_record_relative(paddr_t o) const;
+  paddr_t maybe_relative_to(paddr_t base) const;
+
+  template <typename T>
+  T* cast() {
+    return (static_cast<T*>(this));
+  }
+  template <typename T>
+  const T* cast() const {
+    return (static_cast<const T*>(this));
+  }
 
+  /**
+   * paddr_t::operator-
+   *
+   * Only defined for record_relative paddr_ts.  Yields a
+   * block_relative address.
+   */
+  paddr_t operator-(paddr_t rhs) const;
+  bool operator==(const paddr_t& other) const;
+  bool operator!=(const paddr_t& other) const;
+  bool operator<(const paddr_t& other) const;
+  bool operator<=(const paddr_t& other) const;
+  bool operator>(const paddr_t& other) const;
+  bool operator>=(const paddr_t& other) const;
+
+  DENC(paddr_t, v, p) {
+    DENC_START(1, 1, p);
+    denc(v.dev_addr, p);
+    DENC_FINISH(p);
+  }
+};
+
+struct seg_paddr_t : public paddr_t {
+  static constexpr uint64_t SEG_OFF_MASK = std::numeric_limits<uint32_t>::max();
+  static constexpr uint16_t SEG_OFF_LEN_BITS = 32;
+  // mask for segment manager id
+  static constexpr uint64_t SEG_ID_MASK = 
+    (common_addr_t)0xFFFFFFF << SEG_OFF_LEN_BITS;
+
+  seg_paddr_t(common_addr_t addr) 
+    :paddr_t(addr) {}
```

Comment:
seg_paddr_t probably shouldn't have any actual constructors, the only way to use it should be via the cast<> primitive in paddr_t.